### DISCUSSION TITLE
Managed Node.js scripts

### DISCUSSION THREAD

**Discussion by swkeever** [vercel/next.js#37191] [Ideas]
As far as I know, if you want to create an ad-hoc Node.js script, you have to run it as an ordinary, CommonJS Node script unless you do more setup. See [with-sitemap](https://github.com/vercel/next.js/tree/canary/examples/with-sitemap) for an example of what I mean. 

This has some drawbacks:
- You don't get frictionless ESM support (need to use `.mjs`)
- More setup needed to write TypeScript
- Lose context with the rest of the project (e.g. you can't use your project's module path aliases)

**It would be cool if Next.js managed ad-hoc scripts for you such that we could effortlessly write ESM, TypeScript, and get the same developer experience as when writing application code.**

Here's my idea on how it could work from a developer's perspective. 

In `next.config.js` you could have:
```js
module.exports = {
  scriptsDir: './scripts' // `false` by default, which indicates there are no ad-hoc scripts
}
```
Since `scriptsDir` is `false` by default, that means this feature could be _completely opt-in_.

Here's a contrived example of an ad-hoc script someone might create. Notice we can use ESM, TypeScript, module aliases, and all the DX features we enjoy when writing application code.
```ts
// scripts/seed_db.ts
import db from '@lib/db' // module aliases, ESM syntax

async function seedDatabase(databaseUri?: string) { ... } // TypeScript

await seedDatabase(process.env.DATABASE_URI) // top-level await, environment variables
```

Then perhaps to run the script, you could do something like:
```bash
next script seed_db
```

Perhaps flags could control the behavior further:
```bash
next script seed_db --env=development --watch
```

I find myself often needing to create ad-hoc Node.js scripts when working on a Next.js project, and it's always a bit of a disappointment when I have to fall back to CommonJS and no TypeScript without extra setup. It'd be great if Next.js made doing this simpler. Does anyone else think this would be useful?

**Comment by trevorgerhardt** [comment]
This would be incredibly useful for me as well. If anyone has good workarounds to utilize the Next.js compiler and ts setup for scripts, please share!

**Comment by GabenGar** [comment]
While not completely friction-less you can set the module type on the entire folder [with `package.json`](https://nodejs.org/docs/latest-v16.x/api/packages.html#determining-module-system).
Also there is [`imports` field](https://nodejs.org/docs/latest-v16.x/api/packages.html#subpath-patterns) in `package.json` which has a pretty similar syntax to ts path aliases.
